homeassistant/components/discord/calendar.py

```diff
@@ -0,0 +1,100 @@
+"""Module for using the calendar entity platform."""
+
+from __future__ import annotations
+
+import datetime
+import logging
+from typing import Any
+
+import nextcord
+
+from homeassistant.components.calendar import CalendarEntity, CalendarEvent
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+
+from .const import DOMAIN
+
+PARALLEL_UPDATES = 1
+SCAN_INTERVAL = datetime.timedelta(minutes=1)
+
+_LOGGER = logging.getLogger(__name__)
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    entry: ConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Initialize the calendar entities from the configuration guilds."""
+    client = hass.data[DOMAIN][entry.entry_id]["client"]
+    async_add_entities(
+        [DiscordCalendar(guild) async for guild in client.fetch_guilds()],
+        False,
+    )
+
+
+class DiscordCalendar(CalendarEntity):
+    """A :py:class:`CalendarEntity` subclass that grabs events from discord."""
+
+    def __init__(self, guild: nextcord.Guild) -> None:
+        """Initialize attributes for :py:class:`DiscordCalendar`."""
+        self._guild = guild
+        self._events: list[CalendarEvent] = []
+        self.entity_id = f"calendar.{DOMAIN}_{guild.id}"
+
+    @property
+    def unique_id(self) -> str:
+        """Uniquely identify each calendar with the ID of its corresponding guild."""
+        return str(self._guild.id)
+
+    @property
+    def name(self) -> str:
+        """Display the name of the corresponding guild."""
+        return self._guild.name
+
+    @property
+    def event(self) -> CalendarEvent | None:
+        """Return the oldest event fetched most recently."""
+        if self._events:
+            return self._events[0]
+        return None
+
+    @property
+    def entity_picture(self) -> str:
+        """Return the Bot's profile picture to use in the frontend."""
+        if self._guild.icon is None:
+            return ""
+        return self._guild.icon.url
+
+    @property
+    def extra_state_attributes(self) -> dict[str, Any] | None:
+        """Return extra information about the corresponding guild."""
+        return {"guild.id": str(self._guild.id)}
+
+    async def async_get_events(
+        self,
+        hass: HomeAssistant,
+        start_date: datetime.datetime,
+        end_date: datetime.datetime,
+    ) -> list[CalendarEvent]:
+        """Return all events known locally between two timestamps."""
+        return [
+            event
+            for event in self._events
+            if start_date <= event.start <= event.end <= end_date
```

Comment:
My impression is that start/end time matching should be exclusive: https://developers.home-assistant.io/docs/core/entity/calendar/#get-events

This also only includes events that are entirely contained within the start/end time range and not events that are active during the start/end time, which I think is not exactly correct?

e.g. if you have a trigger, you need to return the event to capture the start/end edge even if the event started or ended outside of the search window.

Reply:
Oh yeah you're right, they are exclusive.

We can access the status of the event as documented on the [Discord Rest API](https://discord.com/developers/docs/resources/guild-scheduled-event#guild-scheduled-event-object-guild-scheduled-event-status) but I don't see it in [nextcord](https://docs.nextcord.dev/en/stable/api.html#nextcord.ScheduledEvent)

Reply:
We can instead check only if the start is between `start_date` and `end_date`

Reply:
I think it's a little more than that really you have these cases to consider:
- start time triggers
- end time triggers
- return active events during the time range

Reply:
So the cases we have.

If the `entity_type` is `EXTERNAL` then the event is active automatically at the scheduled `start_time` and ends automatically at the scheduled `end_time`.

If the `entity_type` is `STAGE_INSTANCE` or `VOICE` then there is a scheduled `start_time` but not a scheduled `end_time` and we assume it is `+timedelta(hours=1)` from the `start_time`. But in most use cases we want to trigger things based on the scheduled times and not whether the event is active or not outside of its scheduled times. Meaning that we don't need to check the status of the event.

Fetching events from discord only returns Active/Scheduled events in the future which suits our purposes.

Reply:
From what I understand from my testing, `async_get_events` is called when a user views it on the frontend. Viewing events in lets says December will request events between December 1st and December 30th. So we need to return events that start between those times exclusively.

For triggering purposes start times are reliable but end times are only reliable when the events are `EXTERNAL` otherwise discord will manage when and where they end based on when all users have left the stage/voice channels which we can't predict and we assume an arbitrary ending time of 1 hour after the start which will warn the users about in the documentation.